src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
-        """Do the work of creating a user's lab.
+    async def delete_lab(self, username: str) -> None:
+        """Delete the lab environment for the given user.
 
-        This method is responsible for creating the Kubernetes objects and
-        telling Kubernetes to start the user's pod. It does not wait for the
-        pod to finish starting. It is run within a background task managed by
-        `~jupyterlabcontroller.services.state.LabStateManager`, which then
-        waits for the lab to start and updates internal state as appropriate.
+        This may be called multiple times for the same user, and all deletions
+        will wait for the same underlying Kubernetes operation.
 
         Parameters
         ----------
-        user
-            Identity information for the user spawning the lab.
-        lab
-            Specification for the lab environment to create.
-        image
-            Docker image to run as the lab.
-        delete_first
-            Whether there is an existing lab that needs to be deleted first.
-
-        Returns
-        -------
-        str
-            Cluster-internal URL at which the lab will be listening once it
-            has finished starting.
+        username
+            Username whose environment should be deleted.
 
         Raises
         ------
-        KubernetesError
-            Raised if there is some failure in a Kubernetes API call.
-        MissingSecretError
-            Raised if a secret does not exist.
+        LabDeletionError
+            Raised if the lab deletion failed for any other reason.
+        OperationConflictError
+            Raised if another operation is already in progress.
+        UnknownUserError
+            Raised if no lab currently exists for this user.
         """
-        username = user.username
+        if username not in self._labs:
+            raise UnknownUserError(f"Unknown user {username}")
+        lab = self._labs[username]
+        if not lab.state:
+            raise UnknownUserError(f"Unknown user {username}")
 
-        # Delete an existing failed lab first if needed.
-        if delete_first:
-            self._logger.info("Deleting existing failed lab")
-            await self._lab_state.publish_event(
-                username, f"Deleting existing failed lab for {username}", 2
+        # There are three possible cases.
+        #
+        # 1. A deletion is already in progress, in which case we want to wait
+        #    on it with everyone else who is monitoring the deletion.
+        # 2. No operation is in progress, in which case we start the deletion
+        #    and take responsibility for clearing the lab state when it
+        #    finishes.
+        # 3. A different operation (presumably a spawn) is in progress, in
+        #    which case we want to raise an error, since currently we don't
+        #    support aborting a spawn in progress by deleting the lab.
+        if lab.monitor.in_progress == _LabOperation.DELETE:
```

Comment:
Can something change the lab.monitor.in_progress state _while_ the if is being tested?  That is, should you assign lab.monitor.in_progress to a temporary variable just before the if, and use that?  Or is it immaterial?

Reply:
The guarantee for async code is that only await can yield control, so as long as you don't await anything, no one else can do things behind your back since only one task is actually running at a time.